From c7b3da0db33803b01f3813797283821072288b94 Mon Sep 17 00:00:00 2001
From: Jakob Perry <japerry@45640.no-reply.drupal.org>
Date: Mon, 22 May 2017 14:04:49 -0400
Subject: [PATCH] Issue 2869592

---
 core/modules/system/system.install | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/core/modules/system/system.install b/core/modules/system/system.install
index 56ce685..5daf6a8 100644
--- a/core/modules/system/system.install
+++ b/core/modules/system/system.install
@@ -817,7 +817,7 @@ function system_requirements($phase) {
     if (!\Drupal::moduleHandler()->moduleExists('update')) {
       $requirements['update status'] = [
         'value' => t('Not enabled'),
-        'severity' => REQUIREMENT_WARNING,
+        'severity' => REQUIREMENT_INFO,
         'description' => t('Update notifications are not enabled. It is <strong>highly recommended</strong> that you enable the Update Manager module from the <a href=":module">module administration page</a> in order to stay up-to-date on new releases. For more information, <a href=":update">Update status handbook page</a>.', [
           ':update' => 'https://www.drupal.org/documentation/modules/update',
           ':module' => \Drupal::url('system.modules_list'),
-- 
2.6.4 (Apple Git-63)

